### PR TITLE
[decoupled-execution] Do not merge: Corner cases and test fixes

### PR DESCRIPTION
**Pull request by yuxiamit**
Do not merge yet

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update or suggest changes to the docs at https://developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: consensus/src/block_storage/sync_manager.rs [hunk 1]
```diff
@@ -295,6 +299,7 @@ impl BlockRetriever {
         peers: &mut Vec<&AccountAddress>,
         num_blocks: u64,
     ) -> anyhow::Result<Vec<Block>> {
+        //info!("Retrieving {} blocks starting from {}", num_blocks, block_id);
```
Comment (zekun000) [inline consensus/src/block_storage/sync_manager.rs:302]: why? is it spamming?

File: consensus/src/epoch_manager.rs [hunk 2]
```diff
@@ -588,10 +601,13 @@ impl EpochManager {
...
             | ConsensusMsg::CommitVoteMsg(_)
             | ConsensusMsg::CommitDecisionMsg(_) => {
-                let event: UnverifiedEvent = msg.into();
+                let event: UnverifiedEvent = msg.clone().into();
                 if event.epoch() == self.epoch() {
                     return Ok(Some(event));
-                } else {
+                } else if !matches!(
+                    msg,
+                    ConsensusMsg::CommitVoteMsg(_) | ConsensusMsg::CommitDecisionMsg(_)
```
Comment (zekun000) [inline consensus/src/epoch_manager.rs:609]: why? these messages carry the epoch number too

File: consensus/src/experimental/commit_phase.rs [hunk 3]
```diff
@@ -141,9 +148,10 @@ pub struct CommitPhase {
     author: Author,
     back_pressure: Arc<AtomicU64>,
     network_sender: NetworkSender,
-    timeout_event_tx: Sender<CommitVote>,
-    timeout_event_rx: Receiver<CommitVote>,
+    timeout_event_tx: Sender<CommitPhaseTimeoutEvent>,
+    timeout_event_rx: Receiver<CommitPhaseTimeoutEvent>,
     reset_event_rx: Receiver<oneshot::Sender<ResetAck>>,
+    in_epoch: bool,
```
Comment (zekun000) [inline consensus/src/experimental/commit_phase.rs:154]: end_epoch?

File: consensus/src/experimental/execution_phase.rs [hunk 4]
```diff
@@ -21,9 +23,17 @@ use std::sync::Arc;
 pub type ResetAck = ();
 pub fn reset_ack_new() -> ResetAck {}
 
+pub type ExecutionPhaseCallBackType = Option<Box<dyn FnOnce(HashValue, Block) + Send + Sync>>;
```
Comment (zekun000) [inline consensus/src/experimental/execution_phase.rs:26]: why this is Option?

File: testsuite/smoke-test/src/storage.rs [hunk 5]
```diff
@@ -284,9 +284,11 @@ fn transfer_and_reconfig(
         if random::<u16>() % 10 == 0 {
             println!(
                 "Changing diem version to {}: {:?}",
-                transferred,
-                client
-                    .change_diem_version(&["change_diem_version", &transferred.to_string()], true)
+                transferred + 1.0,
+                client.change_diem_version(
+                    &["change_diem_version", &(transferred + 1.0).to_string()],
```
Comment (zekun000) [inline testsuite/smoke-test/src/storage.rs:289]: seems wrong we have to change this to pass test

File: config/src/config/consensus_config.rs [hunk 6]
```diff
@@ -48,7 +48,7 @@ impl Default for ConsensusConfig {
             sync_only: false,
             mempool_poll_count: 1,
             decoupled_execution: true, // by default, we turn of the decoupling execution feature
-            channel_size: 30,           // hard-coded
+            channel_size: 60,          // hard-coded
```
[the remark this answers is not in the document]
Comment (yuxiamit) [reply config/src/config/consensus_config.rs:51]: resolved

File: consensus/consensus-types/src/sync_info.rs [hunk 7]
```diff
@@ -191,9 +191,9 @@ impl SyncInfo {
     }
 
     pub fn has_newer_certificates(&self, other: &SyncInfo) -> bool {
-        self.highest_certified_round() > other.highest_certified_round()
+        self.highest_ledger_info_round() > other.highest_ledger_info_round()
```
[the remark this answers is not in the document]
Comment (yuxiamit) [reply consensus/consensus-types/src/sync_info.rs:194]: resolved

File: consensus/src/block_storage/block_store.rs [hunk 8]
```diff
@@ -140,6 +140,41 @@ pub fn update_counters_and_prune_blocks(
...
+}
+
+pub fn path_from_commit_root_with_block_tree(
+    block_id: HashValue,
+    block_tree: Arc<RwLock<BlockTree>>,
+) -> Option<Vec<Arc<ExecutedBlock>>> {
+    block_tree.read().path_from_commit_root(block_id)
+}
+
+pub fn recover_executor_block_tree_and_retry_execute_with_block_tree_and_state_computer(
```
[the remark this answers is not in the document]
Comment (yuxiamit) [reply consensus/src/block_storage/block_store.rs:162]: resolved

File: consensus/src/block_storage/block_store.rs [hunk 9]
```diff
@@ -281,9 +316,12 @@ impl BlockStore {
 
         assert!(!blocks_to_commit.is_empty());
 
-        let block_tree = self.inner.clone();
+        // prepare values for callback
+        let block_tree_in_commit_callbacl = self.inner.clone();
```
[the remark this answers is not in the document]
Comment (yuxiamit) [reply consensus/src/block_storage/block_store.rs:320]: resolved

File: consensus/src/block_storage/sync_manager.rs (hunk 1 above)
Comment (yuxiamit) [reply consensus/src/block_storage/sync_manager.rs:302]: yes, it spams quite a bit before we changed the limit

Comment (yuxiamit) [reply consensus/src/block_storage/sync_manager.rs:302]: Yeah a little bit while I was debugging other stuff.

File: consensus/src/epoch_manager.rs (hunk 2 above)
Comment (yuxiamit) [reply consensus/src/epoch_manager.rs:609]: previously we had the issue of hanging around commit phase, who keeps sending stale epoch messages, causing the node to send to itself useless epoch change proofs.

File: consensus/src/experimental/commit_phase.rs (hunk 3 above)
Comment (yuxiamit) [reply consensus/src/experimental/commit_phase.rs:154]: it means whether the commit phase is in the epoch (the contrary to end_epoch).

File: consensus/src/experimental/execution_phase.rs (hunk 4 above)
Comment (yuxiamit) [reply consensus/src/experimental/execution_phase.rs:26]: To make the test cleaner.

File: consensus/src/epoch_manager.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/epoch_manager.rs:609]: but it shouldn't happen after we fix the commit phase?

File: testsuite/smoke-test/src/storage.rs (hunk 5 above)
Comment (yuxiamit) [reply testsuite/smoke-test/src/storage.rs:289]: do you mean the test is wrong?

File: consensus/src/experimental/commit_phase.rs (hunk 3 above)
Comment (zekun000) [reply consensus/src/experimental/commit_phase.rs:154]: `in_epoch` is a little confusing, since it should always be in an epoch. we can flip the bool and name is as `epoch_end` or simply `stop`

File: consensus/src/epoch_manager.rs (hunk 2 above)
Comment (yuxiamit) [reply consensus/src/epoch_manager.rs:609]: right, but the current commit phase fix is not clean either. We can leave it here, since it is okay to ignore commit votes and decisions with wrong epoch.

File: testsuite/smoke-test/src/storage.rs (hunk 5 above)
Comment (zekun000) [reply testsuite/smoke-test/src/storage.rs:289]: no, if the test passes today, it should pass without changes.

File: consensus/src/experimental/execution_phase.rs (hunk 4 above)
Comment (zekun000) [reply consensus/src/experimental/execution_phase.rs:26]: we shouldn't adopt code for test, can't we just create empty callback?

File: consensus/src/epoch_manager.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/epoch_manager.rs:609]: this looks very inconsistent with other messages

File: consensus/src/experimental/execution_phase.rs (hunk 4 above)
Comment (yuxiamit) [reply consensus/src/experimental/execution_phase.rs:26]: Okay, I'll change this a bit later.

File: consensus/src/experimental/commit_phase.rs (hunk 3 above)
Comment (yuxiamit) [reply consensus/src/experimental/commit_phase.rs:154]: okay I'll change it later

File: consensus/src/epoch_manager.rs (hunk 2 above)
Comment (yuxiamit) [reply consensus/src/epoch_manager.rs:609]: I think it has a reason as these messages are asynchronous and should not involve in the ordering?

Comment (zekun000) [reply consensus/src/epoch_manager.rs:609]: this is not about ordering, it's about epoch which is related to all phases.
